File: 이훈일/Programmers/Level2/롤케이크 자르기.java
import java.util.*;

class Solution {
    public int solution(int[] topping) {
        HashMap<Integer, Integer> first = new HashMap<>();
        HashMap<Integer, Integer> second = new HashMap<>();
        
        for(int t : topping) {
            first.put(t, first.getOrDefault(t, 0) + 1);
        }
        
        int count = 0;
        for(int t : topping) {
            second.put(t, second.getOrDefault(t, 0) + 1);
            
            if(first.get(t) == 1) {
                first.remove(t);
            } else {
                first.put(t, first.get(t) - 1);
            }
            
            if(first.size() == second.size()) {
                count++;
            }
        }
        
        return count;
    }
}

File: 조준장/롤케이크 자르기.kt
class Solution {
    fun solution(topping: IntArray): Int {
        var answer: Int = 0
        var a = HashMap<Int, Int>()
        var b = mutableSetOf<Int>()
        topping.toSet().forEach{ num ->
            a += num to topping.count{it == num}
        }
        
        topping.forEach{ tp ->
            a[tp] = a[tp]!! - 1
            b.add(tp)
            
            if (a[tp] == 0) a.remove(tp)
            if (a.size == b.size) answer++
    
        }
        
        
        return answer
    }
}
